src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs

```diff
@@ -117,37 +117,33 @@ public void EnableMeasurementEvents(Instrument instrument, object? state = null)
         /// <param name="measurementCallback">The callback which can be used to get measurement recording of numeric type T.</param>
         public void SetMeasurementEventCallback<T>(MeasurementCallback<T>? measurementCallback) where T : struct
         {
-            if (measurementCallback is MeasurementCallback<byte> byteCallback)
+            if (measurementCallback is null)
+                measurementCallback = (instrument, measurement, tags, state) => { /* no-op */};
+            switch (typeof(T))
             {
-                _byteMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : byteCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<int> intCallback)
-            {
-                _intMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : intCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<float> floatCallback)
-            {
-                _floatMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : floatCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<double> doubleCallback)
-            {
-                _doubleMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : doubleCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<decimal> decimalCallback)
-            {
-                _decimalMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : decimalCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<short> shortCallback)
-            {
-                _shortMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : shortCallback;
-            }
-            else if (measurementCallback is MeasurementCallback<long> longCallback)
-            {
-                _longMeasurementCallback = (measurementCallback is null) ? ((instrument, measurement, tags, state) => { /* no-op */}) : longCallback;
-            }
-            else
-            {
-                throw new InvalidOperationException(SR.Format(SR.UnsupportedType, typeof(T)));
+                case var type when type == typeof(byte):
```

Comment:
Please revert the code as originally written. This is written this way to have the JIT optimize it.

Reply:
`case` is worst than `if` for JIT? I did not know this, Did you have any links to I can read more about that?

thanks

Reply:
I don't think the JIT optimization is documented. This is something internal which can change.

Reply:
Ok, I change it, but It is very strange, because usually `switch case` is faster statement than `if else if` in another language, for example C++

Reply:
JIT optimization is actually removing the if conditions totally. JIT knows the generic type and eliminates all these conditions :-) isn't nice?

Reply:
It is very nice :) , but I expect JIT is same (or better) behavior for `switch case` statement with `pattern`